File: BinaryTree/binarytree.cc
#include <stdlib.h>
#include <stdio.h>

struct TreeNode
{
	int value;
	struct TreeNode *left;
	struct TreeNode *right;
};

typedef struct TreeNode* Node;

Node Insert(Node root,int val)
{
	if(root == NULL)
	{
		Node temp;
		temp = (struct TreeNode *)malloc(sizeof(struct TreeNode));
		temp->value = val;
		temp->left = NULL;
		temp->right = NULL;
		return temp;
	}
	if((root->value == 0 && val == 1) || (root->value == 1 && val == 0))
		root->right =  Insert(root->right,val);
	else
		root->left =  Insert(root->left,val);
	return root;
}

void PrintTree(Node root)
{
	if(!root)
		return;
	PrintTree(root->left);
	printf("%d",root->value);
	PrintTree(root->right);
}

void DestroyTree(Node root)
{
	if(root)
	{
		DestroyTree(root->left);
		DestroyTree(root->right);
		free(root);
	}
}

int main()
{
	Node root;
	root = (struct TreeNode *)malloc(sizeof(struct TreeNode));
	root->value = 0;
	root->left = NULL;
	root->right = NULL;
	root = Insert(root,0);
	root = Insert(root,1);
	root = Insert(root,1);
	root = Insert(root,0);
	root = Insert(root,1);
	root = Insert(root,0);
	root = Insert(root,1);
	root = Insert(root,1);
	root = Insert(root,0);
	PrintTree(root);

	DestroyTree(root);
	return 0;
}

File: Problems/src/problems/BinaryTree.java
package problems;

import java.util.*;
import java.util.LinkedList;

class TreeNode
{

	public int data;
	public TreeNode left = null;
	public TreeNode right = null;

	public TreeNode(int d)
	{
		data = d;
	}
}

public class BinaryTree
{
	TreeNode root;

	public void deleteTree()
	{
		root = null;
	}

	public int height(TreeNode root)
	{
		if (root == null)
			return 0;

		return Math.max(height(root.left), height(root.right)) + 1;
	}

	public void printInOrder(TreeNode root)
	{
		if (root == null)
			return;
		printInOrder(root.left);
		System.out.print(root.data + " ");
		printInOrder(root.right);
	}
	
	static TreeNode last_node = null;
	public TreeNode inOrderSuccessor(TreeNode root, TreeNode node)
	{
		if (root == null || node == null)
			return null;
		TreeNode succ = null;

		succ = inOrderSuccessor(root.left, node);
		if (last_node != null && last_node == node)
			return root;
		last_node = root;
		succ = inOrderSuccessor(root.right, node);

		return succ;
	}
	
	public TreeNode find(TreeNode root, int d)
	{
		if (root == null)
			return null;

		if (root.data == d)
			return root;

		TreeNode temp;
		
		temp = find(root.left,d);
		if(temp != null)
			return temp;
		temp = find(root.right,d);
		return temp;
		
	}
	
	public TreeNode insert(TreeNode root, int d)
	{
		return null;
	}
	
	public void printPaths()
	{
		int[] path = new int[height(root)];
		printPaths(root, path, 0);
	}

	private void printPaths(TreeNode node, int[] path, int level)
	{
		if (node == null)
			return;

		path[level] = node.data;
		//System.out.println("level: " + level);
		
		if (node.left == null && node.right == null)
		{
			printArray(path, level);
		}
		else
		{
			printPaths(node.left, path, level+1);
			printPaths(node.right, path, level+1);
		}
	}

	private void printArray(int[] ints, int level)
	{
		for (int i = 0; i <= level; i++)
			System.out.print(ints[i] + " ");
		System.out.println();
	}
	
	public boolean BFS(TreeNode root, int d)
	{
		if(root == null) return false;
		
		if(root.data == d) return true;
		
		Queue<TreeNode> q = new LinkedList<TreeNode>();
		TreeNode temp;
		q.offer(root);
		
		while(!q.isEmpty())
		{
			 temp = q.poll();
			 if(temp != null)
			 {
				 if(temp.data == d)
					 return true;
    			 if(temp.left != null)
    				 q.offer(temp.left);
    			 if(temp.right != null)
    				 q.offer(temp.right);
			 }
		}
		return false;
	}
	
	public boolean DFS(TreeNode root, int d)
	{
		if(root == null) return false;
		
		if(root.data == d) return true;
		
		return (DFS(root.left, d) || DFS(root.right,d));
	}
	
}
